### PR TITLE
parser: Ensure that all nonterminals have tokens after parsing

### PR DESCRIPTION
**Pull request by petrochenkov**
`parse_nonterminal` should always result in something with tokens.

This requirement wasn't satisfied in two cases:
- `stmt` nonterminal with expression statements (e.g. `0`, or `{}`, or `path + 1`) because `fn parse_stmt_without_recovery` forgot to propagate `force_collect` in some cases.
- `expr` nonterminal with expressions with built-in attributes (e.g. `#[allow(warnings)] 0`) due to an incorrect optimization in `fn parse_expr_force_collect`, it assumed that all expressions starting with `#` have their tokens collected during parsing, but that's not true if all the attributes on that expression are built-in and inert.

(Discovered when trying to implement eager `cfg` expansion for all attributes https://github.com/rust-lang/rust/pull/83824#issuecomment-817317170.)

r? @Aaron1011

### CODE REVIEW

File: compiler/rustc_parse/src/parser/attr_wrapper.rs [hunk 1]
```diff
@@ -342,16 +342,10 @@ impl<'a> Parser<'a> {
 
         // If we support tokens at all
         if let Some(target_tokens) = ret.tokens_mut() {
-            if let Some(target_tokens) = target_tokens {
-                assert!(
```
Comment (Aaron1011) [inline compiler/rustc_parse/src/parser/attr_wrapper.rs:346]: Why did you remove this assertion?

Comment (petrochenkov) [reply compiler/rustc_parse/src/parser/attr_wrapper.rs:346]: It's hit in cases like `#[cfg_eval] #[rustc_dummy] 0`.

Tokens for `#[rustc_dummy] 0` are collected twice due to the new behavior of `parse_expr_force_collect`, but we still get to here despite the
```rust
        if !self.capture_cfg && matches!(ret.tokens_mut(), None | Some(Some(_))) {
            return Ok(ret);
        }
```
above because in case of `cfg_eval` the `self.capture_cfg` condition is true.

File: src/test/ui/macros/expr-stmt-nonterminal-tokens.rs [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+
+macro_rules! mac {
+    (expr $expr:expr) => {};
+    (stmt $stmt:stmt) => {};
+}
+
+fn main() {
+    mac!(expr #[allow(warnings)] 0);
+    mac!(stmt 0);
+    mac!(stmt {});
```
Comment (petrochenkov) [inline src/test/ui/macros/expr-stmt-nonterminal-tokens.rs:11]: This `libc` failure on CI is
```rust
mac!(stmt path + 1);
```

File: compiler/rustc_parse/src/parser/nonterminal.rs [hunk 3]
```diff
@@ -169,7 +170,19 @@ impl<'a> Parser<'a> {
                     return Err(self.struct_span_err(self.token.span, msg));
                 }
             }
-        })
+        };
+
+        // If tokens are supported at all, they should be collected.
+        if !matches!(nt.tokens_mut(), None | Some(Some(..))) {
```
Comment (Aaron1011) [inline compiler/rustc_parse/src/parser/nonterminal.rs:176]: I think this could be written more clearly as `matches!(nt.tokens_mut(), Some(None)`

File: src/test/ui/macros/expr-stmt-nonterminal-tokens.rs [hunk 4]
```diff
@@ -0,0 +1,15 @@
+// check-pass
```
Comment (Aaron1011) [inline src/test/ui/macros/expr-stmt-nonterminal-tokens.rs:1]: This file compiles successfully on the latest nightly. I think you need to force recollection of the nonterminals by a proc-macro in order to trigger the panic.

Comment (petrochenkov) [reply src/test/ui/macros/expr-stmt-nonterminal-tokens.rs:1]: Yes, the test assumes existence of the missing tokens panic added in this PR, it won't fail without it.

**Comment by Aaron1011** [comment]
Let's see what the performance impact of disabling the expr optimization looks like:

@bors try @rust-timer queue

**Comment by petrochenkov** [comment]
Minor regression.
Not sure why exactly, `stmt` matchers and `cfg_eval` are exceedingly rare, and expressions with attributes in `expr` matchers should be rare as well.

File: compiler/rustc_parse/src/parser/stmt.rs [hunk 5]
```diff
@@ -85,21 +89,22 @@ impl<'a> Parser<'a> {
             self.mk_stmt(lo, StmtKind::Empty)
         } else if self.token != token::CloseDelim(token::Brace) {
             // Remainder are line-expr stmts.
-            let e = self.parse_expr_res(Restrictions::STMT_EXPR, Some(attrs))?;
+            let e = if force_collect == ForceCollect::Yes {
+                self.collect_tokens_no_attrs(|this| {
```
Comment (Aaron1011) [inline compiler/rustc_parse/src/parser/stmt.rs:93]: This should be a call to `collect_tokens_trailing_token`, with `attrs` passed in (since we do have outer attributes).

File: compiler/rustc_parse/src/parser/expr.rs [hunk 6]
```diff
@@ -94,17 +94,7 @@ impl<'a> Parser<'a> {
...
-        if matches!(self.token.kind, TokenKind::Pound | TokenKind::DocComment(..)) {
-            self.parse_expr()
-        } else {
-            // If we don't have outer attributes, then we need to ensure
-            // that collection happens by using `collect_tokens_no_attrs`.
-            // Expression don't support custom inner attributes, so `parse_expr`
-            // will never try to collect tokens if we don't have outer attributes.
-            self.collect_tokens_no_attrs(|this| this.parse_expr())
-        }
+        self.collect_tokens_no_attrs(|this| this.parse_expr())
```
Comment (Aaron1011) [inline compiler/rustc_parse/src/parser/expr.rs:97]: Normally, this would be incorrect - `collect_tokens_no_attrs` was only supposed to be used when the target was guaranteed to not have any outer attribute - if outer attributes are present, we need to capture their tokens separately so that we can create a proper `ReplaceRange`.

However, we only actually need to know the attribute start position when we're in `capture_cg` mode, which never happens when we're parsing nonterminals (we flatten all nonterminals before entering `capture_cfg` mode). Can you add a `debug_assert!(!self.capture_cfg)` to `parse_expr_force_collect` to make sure this method doesn't get misused?

I may do some refactoring in a follow-up PR.

Comment (petrochenkov) [reply compiler/rustc_parse/src/parser/expr.rs:97]: The assert will fire because `parse_expr_force_collect` is indeed used with `capture_cfg == true`.
Here - https://github.com/rust-lang/rust/blob/master/compiler/rustc_builtin_macros/src/cfg_eval.rs#L194

(The result of `cfg_eval` should have tokens collected, especially with #85073.)

Comment (petrochenkov) [reply compiler/rustc_parse/src/parser/expr.rs:97]: I assumed that wrapping into `collect_tokens_no_attrs` is ok because there are two cases:
- There are active attributes, then token collection will be called once again in `parse_expr`, will set tokens for the returned expression, so the tokens collected by `collect_tokens_no_attrs` will simply be thrown away.
(I'm not sure what happens with range replacement in this case.)
- There are no active attributes, then `collect_tokens_no_attrs` is actually doing the right thing and it will set tokens for the returned expression. (I'm again not sure what will happen with range replacement in this case.)

File: compiler/rustc_parse/src/parser/stmt.rs (hunk 5 above)
Comment (petrochenkov) [reply compiler/rustc_parse/src/parser/stmt.rs:93]: I assumed this case is analogous to `parse_expr_force_collect` (https://github.com/rust-lang/rust/pull/84995/files#r629329546) so we can use `collect_tokens_no_attrs`.

Also, I don't think we can simply pass `attrs` here because they don't necessary apply to the whole statement, they can apply only to some sub-expression like in `#[attr] 1 + 2` (grouped as `(#[attr] 1) + 2`).

File: src/test/ui/macros/expr-stmt-nonterminal-tokens.rs [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+
+macro_rules! mac {
```
Comment (Aaron1011) [inline src/test/ui/macros/expr-stmt-nonterminal-tokens.rs:3]: Can you modify this test (or add a test) to expand to `#[derive]` target, which verifies that the tokens from the nonterminals are cfg-expanded as expected. E.g.:

```rust
macro_rules! mac {
    (expr $expr:expr) => {
        #[derive(Print)]
        struct Foo {
            field: [bool; { let  = $expr; 0 }]
        }
    }
}
```

File: compiler/rustc_parse/src/parser/expr.rs (hunk 6 above)
Comment (Aaron1011) [reply compiler/rustc_parse/src/parser/expr.rs:97]: I think it will happen to do the right thing with replace ranges - we'll end up creating a `ReplaceRange` covering the entire expression (including outer attributes), as expected. Since we'll throw away the new tokens, the final captured tokens will *not* include the outer attributes, which will allow us to correctly perform cfg-expansion.

I forgot about the fact that we use force collection at the 'top level' of `cfg_eval`. It's unfortunate that we'll lose the asertion, but I don't think keeping it is worth the extra complexity of keeping track of the 'top-level' status on top of everything else.

File: src/test/ui/macros/expr-stmt-nonterminal-tokens.rs [hunk 8]
```diff
@@ -0,0 +1,15 @@
...
+    (expr $expr:expr) => {};
+    (stmt $stmt:stmt) => {};
+}
+
+fn main() {
+    mac!(expr #[allow(warnings)] 0);
```
[the remark this answers is not in the document]
Comment (petrochenkov) [reply src/test/ui/macros/expr-stmt-nonterminal-tokens.rs:9]: Feature-gating for expression/statement attributes has many holes.
I tried fixing them but it causes breakage in practice unfortunately.
So I thought perhaps it's better to work in the direction of the stabilization instead, which should cause less breakage.

File: src/test/ui/proc-macro/expr-stmt-nonterminal-tokens.stdout [hunk 9]
```diff
@@ -0,0 +1,540 @@
+PRINT-DERIVE INPUT (DISPLAY): enum E { V = { let _ = #[allow(warnings)] 0 ; 0 }, }
+PRINT-DERIVE DEEP-RE-COLLECTED (DISPLAY): enum E { V = { let _ = #[allow(warnings)] #[allow(warnings)] 0 ; 0 }, }
```
Comment (petrochenkov) [inline src/test/ui/proc-macro/expr-stmt-nonterminal-tokens.stdout:2]: @Aaron1011 
The attribute in `$expr` `#[allow(warnings)] 0` is duplicated for some reason.

Comment (petrochenkov) [reply src/test/ui/proc-macro/expr-stmt-nonterminal-tokens.stdout:2]: The duplication reproduces with
```rust
macro mac($expr:expr) {
    print_bang!($expr);
}

fn main() {
    mac!(#[allow(warnings)] 0);
}
```
as well.

At least this is not a regression, previously these examples resulted in ICEs.

Comment (petrochenkov) [reply src/test/ui/proc-macro/expr-stmt-nonterminal-tokens.stdout:2]: Made #86055 for tracking this.
